### PR TITLE
[DESK] Fix screensaver preview drawing

### PR DESCRIPTION
**Pull request by binarymaster**
Use window subclassing to override `WM_PAINT` message handling and use `RedrawWindow` function along with `WS_CLIPCHILDREN` style for the parent window in order to preserve screensaver drawing. This is what Windows' desk.cpl (themeui.dll) seems to do in order to solve this problem.

Also reuse the same monitor bitmap for the top preview control.

I've split it into two distinct commits for easier review.

JIRA issues: [CORE-15929](https://jira.reactos.org/browse/CORE-15929) and [CORE-10606](https://jira.reactos.org/browse/CORE-10606)

## Demo

Before and after video tested on Windows XP:

https://user-images.githubusercontent.com/578406/193632971-6acb680a-018c-4e69-9e63-5d10d20f42e1.mp4

### CODE REVIEW

File: dll/cpl/desk/lang/bg-BG.rc [hunk 1]
```diff
@@ -33,7 +33,7 @@ STYLE DS_SHELLFONT | WS_CHILD | WS_DISABLED | WS_CAPTION
 CAPTION "Екранен предпазител"
 FONT 8, "MS Shell Dlg", 0, 0
 BEGIN
-    CONTROL "", IDC_SCREENS_PREVIEW, "Static", SS_OWNERDRAW, 70, 10, 105, 70, WS_EX_STATICEDGE
```
Comment (HBelusca) [inline dll/cpl/desk/lang/bg-BG.rc:36]: Is this needed then, the trailing "zero" for the style? (does it still work without it?)

File: dll/cpl/desk/screensaver.c [hunk 2]
```diff
@@ -12,6 +12,8 @@
 
 #define MAX_SCREENSAVERS 100
 
+static const TCHAR szPreviewWndClass[] = TEXT("SSDemoParent");
```
Comment (HBelusca) [inline dll/cpl/desk/screensaver.c:15]: Windows-compatible name?

File: dll/cpl/desk/screensaver.c [hunk 3]
```diff
@@ -594,6 +667,40 @@ OnInitDialog(HWND hwndDlg, PDATA pData)
         return FALSE;
     }
 
+    wc.lpfnWndProc = DefWindowProc;
+    wc.hInstance = hApplet;
+    wc.hCursor = NULL;
+    wc.hbrBackground = (HBRUSH)GetStockObject(BLACK_BRUSH);
+    wc.lpszClassName = szPreviewWndClass;
+
+    if (RegisterClass(&wc))
```
Comment (HBelusca) [inline dll/cpl/desk/screensaver.c:676]: Better use RegisterClassEx with a WNDCLASSEX.

File: dll/cpl/desk/lang/bg-BG.rc (hunk 1 above)
Comment (binarymaster) [reply dll/cpl/desk/lang/bg-BG.rc:36]: I just copypasted an existing control. But yes, it seems to compile fine without it.

File: dll/cpl/desk/screensaver.c (hunk 2 above)
Comment (binarymaster) [reply dll/cpl/desk/screensaver.c:15]: Yep, checked with Spy++.

File: dll/cpl/desk/screensaver.c (hunk 3 above)
Comment (binarymaster) [reply dll/cpl/desk/screensaver.c:676]: Well [MSDN says](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-registerclassa) that "You can still use RegisterClass, however, if you do not need to set the class small icon." -- also this is already used for multimon control here: https://git.reactos.org/?p=reactos.git;a=blob;f=dll/cpl/desk/monslctl.c;hb=3aa5e8b897d4962080ecc765ad6b1aa50d6d886a#l1637

File: dll/cpl/desk/lang/bg-BG.rc (hunk 1 above)
Comment (binarymaster) [reply dll/cpl/desk/lang/bg-BG.rc:36]: Fixed the styles in master and here.

File: dll/cpl/desk/screensaver.c [hunk 4]
```diff
@@ -594,6 +667,40 @@ OnInitDialog(HWND hwndDlg, PDATA pData)
...
+    {
+        HWND hParent = GetDlgItem(hwndDlg, IDC_SCREENS_PREVIEW);
+        HWND hChild;
+
+        if (hParent != INVALID_HANDLE_VALUE)
```
Comment (katahiromz) [inline dll/cpl/desk/screensaver.c:681]: I think `INVALID_HANDLE_VALUE` is provided for kernel handles.

```suggestion
        if (hParent != NULL)
```

File: dll/cpl/desk/screensaver.c [hunk 5]
```diff
@@ -594,6 +667,40 @@ OnInitDialog(HWND hwndDlg, PDATA pData)
...
+            pData->OldPreviewProc = (WNDPROC)GetWindowLongPtr(hParent, GWLP_WNDPROC);
+            SetWindowLongPtr(hParent, GWLP_WNDPROC, (LONG_PTR)RedrawSubclassProc);
+            SetWindowLongPtr(hParent, GWLP_USERDATA, (LONG_PTR)pData);
+        }
+
+        hChild = CreateWindowEx(0, szPreviewWndClass, NULL,
+                                WS_CHILD | WS_CLIPCHILDREN,
+                                0, 0, 0, 0, hParent,
+                                NULL, hApplet, NULL);
+        if (hChild != INVALID_HANDLE_VALUE)
```
Comment (katahiromz) [inline dll/cpl/desk/screensaver.c:692]:
```suggestion
        if (hChild != NULL)
```

File: dll/cpl/desk/screensaver.c [hunk 6]
```diff
@@ -594,6 +667,40 @@ OnInitDialog(HWND hwndDlg, PDATA pData)
         return FALSE;
     }
 
+    wc.lpfnWndProc = DefWindowProc;
+    wc.hInstance = hApplet;
+    wc.hCursor = NULL;
```
Comment (katahiromz) [inline dll/cpl/desk/screensaver.c:672]: `hCursor` is NULL?  Alright?

Comment (binarymaster) [reply dll/cpl/desk/screensaver.c:672]: Yes, seems to work fine on both Windows/ReactOS, also: https://git.reactos.org/?p=reactos.git;a=blob;f=modules/rostests/tests/subclass/subclass.c;hb=826bd41d88df8b83494ce921f7f4a3ec55544c7f#l16

MSDN: https://docs.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-wndclassa

> If this member is NULL, an application must explicitly set the cursor shape whenever the mouse moves into the application's window.

File: dll/cpl/desk/screensaver.c (hunk 4 above)
Comment (binarymaster) [reply dll/cpl/desk/screensaver.c]: Fixed.

**Review by katahiromz — APPROVED** [review]

File: dll/cpl/desk/screensaver.c [hunk 7]
```diff
@@ -727,6 +843,16 @@ ScreenSaverPageProc(HWND hwndDlg,
             break;
         }
 
+        case WM_DRAWITEM:
+        {
+            LPDRAWITEMSTRUCT lpDrawItem;
+            lpDrawItem = (LPDRAWITEMSTRUCT) lParam;
```
Comment (HBelusca) [inline dll/cpl/desk/screensaver.c:849]:
```suggestion
            lpDrawItem = (LPDRAWITEMSTRUCT)lParam;
```

**Review by HBelusca — APPROVED** [review]

**Pull request merged**